### PR TITLE
Implement RuntimeHelpers.PrepareMethod/PrepareDelegate for CoreCLR

### LINKED ISSUE #15522: RuntimeHelpers.PrepareDelegate/PrepareMethod should JIT the given method
There is a comment here:

https://github.com/dotnet/coreclr/blob/b9affb49f79ed8319923c2f5b07b4aeeb04efe36/src/mscorlib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs#L147-L149

But I can't figure out if it is supported or not.

In case it isn't supported, is there another way to ensure certain methods in a call graph are jitted? Or to jit a method without calling it?

### PR DESCRIPTION
**Pull request by jkotas**
CoreCLR implementation of this method triggers jiting of the given method only.
It does not walk a subset of callgraph to provide CER guarantees because of CERs
are not supported by CoreCLR.

Fixes #15522

### CODE REVIEW

File: src/mscorlib/Resources/Strings.resx [hunk 1]
```diff
@@ -3709,4 +3709,13 @@
   <data name="Arg_NullArgumentNullRef" xml:space="preserve">
     <value>The method was called with a null array argument.</value>
   </data>
-</root>
\ No newline at end of file
+  <data name="Argument_CannotPrepareAbstract" xml:space="preserve">
+    <value>Abstract methods cannot be prepared.</value>
+  </data>
+  <data name="Argument_CannotPrepareAbstract" xml:space="preserve">
```
Comment (ilmax) [inline src/mscorlib/Resources/Strings.resx]: This is duplicate, is it intended?

**Comment by jkotas** [comment]
> For my own edification, in what situations are these guarantees (JIT target only) valuable?

The side-effects of `RuntimeHelpers.PrepareMethod` are frequently used for custom implementations of background JIT. There are many articles out there about that:

https://stackoverflow.com/questions/26590809/disadvantages-of-runtimehelpers-preparemethod-in-a-windows-service
https://stackoverflow.com/questions/13435641/questions-about-pre-jit-ing-using-runtimehelpers-preparemethod

The JIT target only implementation is meant to make these kind of uses work. 

I do not think it make sense to try to bring CERs and the partial call-graph preparation to .NET Core. I think that killing the process is much simpler and more robust way to deal with hard to recover from errors. https://github.com/dotnet/coreclr/issues/15522#issuecomment-352031215 has some more context.

File: src/mscorlib/Resources/Strings.resx (hunk 1 above)
Comment (jkotas) [reply src/mscorlib/Resources/Strings.resx]: Fixed. Thanks!

**Comment by stephentoub** [comment]
> frequently used for custom implementations of background JIT

Makes sense. Thanks.

**Comment by AndyAyersMS** [comment]
IIRC, Desktop implementation also would mark any method passed to `PrepareMethod` as noinline. We don'd do that here, and I think that's the right decision. But it may be worth a short discussion.

It's somewhat pointless to forcibly jit a method that will be subsequently be inlined. Marking prepared methods as noinline ensures no further jit resources are expended on the code in these methods. So from the standpoint of precisely controlling when jit costs might be incurred, marking makes sense.

But there's no practical way to limit which methods can be passed to `PrepareMethod`. And calling prepare method on methods with aggressive inline attributes will lead to some surprising behavior, as they are no longer inlined anywhere, for reasons that may be hard to discover. Also marking methods as noinline makes it harder to leverage `PrepareMethod` for jit testing.

So I like the current behavior, even if we'll still sometimes spend time jitting code for methods that were previously prepared, when such methods are inlined.

**Comment by jkotas** [comment]
> Marking prepared methods as noinline ensures no further jit resources are expended on the code in these methods.

Note that marking prepared methods as noinline was not meant as a performance optimization in .NET Framework. It is done to make a subtle details of CERs work. CERs requires the roots to show up on the stack.

**Pull request merged**